src/repositories/github.rs

```diff
@@ -0,0 +1,244 @@
+                last_activity_at: repo.pushed_at,
+                stars: repo.stargazer_count,
+                forks: repo.fork_count,
+                issues: repo.issues.total_count,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn fetch_repositories(&self, node_ids: &[String]) -> Result<FetchRepositoriesResult> {
+        let response: GraphResponse<GraphNodes<Option<GraphRepository>>> = self.graphql(
+            GRAPHQL_UPDATE,
+            serde_json::json!({
+                "ids": node_ids,
+            }),
+        )?;
+
+        // The error is returned *before* we reach the rate limit, to ensure we always have an
+        // amount of API calls we can make at any time.
+        trace!(
+            "GitHub GraphQL rate limit remaining: {}",
+            response.data.rate_limit.remaining
+        );
+        if response.data.rate_limit.remaining < self.github_updater_min_rate_limit {
+            return Err(RateLimitReached.into());
+        }
+
+        let mut ret = FetchRepositoriesResult::default();
+
+        for error in &response.errors {
+            use GraphErrorPath::*;
+            match (error.error_type.as_str(), error.path.as_slice()) {
+                ("NOT_FOUND", [Segment(nodes), Index(idx)]) if nodes == "nodes" => {
+                    ret.missing.push(node_ids[*idx as usize].clone());
+                }
+                _ => failure::bail!("error updating repositories: {}", error.message),
+            }
+        }
+        // When a node is missing (for example if the repository was deleted or made private) the
+        // GraphQL API will return *both* a `null` instead of the data in the nodes list and a
+        // `NOT_FOUND` error in the errors list.
+        for node in response.data.nodes.into_iter() {
+            if let Some(node) = node {
+                let repo = Repository {
+                    id: node.id,
+                    name_with_owner: node.name_with_owner,
+                    description: node.description,
+                    last_activity_at: node.pushed_at,
+                    stars: node.stargazer_count,
+                    forks: node.fork_count,
+                    issues: node.issues.total_count,
+                };
+                ret.present.insert(repo.id.clone(), repo);
+            }
+        }
+
+        Ok(ret)
+    }
+}
+
+impl GitHub {
+    fn graphql<T: serde::de::DeserializeOwned>(
+        &self,
+        query: &str,
+        variables: impl serde::Serialize,
+    ) -> Result<GraphResponse<T>> {
+        Ok(self
+            .client
+            .post("https://api.github.com/graphql")
+            .json(&serde_json::json!({
+                "query": query,
+                "variables": variables,
+            }))
+            .send()?
+            .error_for_status()?
+            .json()?)
+    }
+}
+
+#[derive(Debug, failure::Fail)]
+#[fail(display = "rate limit reached")]
+struct RateLimitReached;
```

Comment:
This should be defined in `src/repositories/updater.rs`, as GitLab should too report when the rate limit is reached.